File: Week4_Exercise/MyObjects/Product.cs
namespace Week4_Exercise.MyObjects;

public abstract class Product
{
    private string _name = string.Empty;
    private int _price;
    private int _quantity;

    public Product()
    {
        _name = "Name not Provided";
        _price = 0;
        _quantity = 0;
    }

    public Product(string name, int price, int quantity)
    {
        _name = name;
        _price = price;
        _quantity = quantity;   
    }

    public string Name 
    {
        get => _name;
        set => _name = value;
    }
    public int Price 
    {
        get => _price;
        set => _price = value;
    }
    public int Quantity 
    {
        get => _quantity;
        set => _quantity = value;
    }

    public abstract void PrintInfo();
    
}

File: Week8_Cpp/Product.cpp
//
// Created by billy on 3/11/2022.
//

#include "Product.h"

#include <utility>
using namespace std;

Product::Product() {
    _prodName = "Unnamed Product";
    _price = 0.00;
    _quantity = 0;

}

Product::Product(string name, double price, int quantity)
        : _prodName(std::move(name)), _price(price), _quantity(quantity){

}

string Product::getName() {
    return _prodName;
}

double Product::getPrice() const {
    return _price;
}

int Product::getQuantity() const {
    return _quantity;
}

double Product::CalculatePrice() {
    return 0.1 * _price;
}

void Product::PrintDetails() {
    cout
            << "this is the base function" << endl;
}
